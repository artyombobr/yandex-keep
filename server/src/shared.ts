export interface NotesProps {
  data: DataEntity;
}

export interface DataEntity {
  tags: TagsEntity[];
  colors: ColorsEntity[];
  notes: NotesEntity[];
}

export interface TagsEntity {
  id: number;
  tag: string;
}

export interface ColorsEntity {
  id: number;
  color: string;
}

export interface NotesEntity {
  type: string;
  title?: string;
  tags?: number[];
  color?: number;
  items?: ItemsEntity[];
  size?: string;
  created: number;
  text?: string;
  attachments?: AttachmentsEntity[];
  reminder?: number;
  url?: string;
  archive?: boolean;
}

export interface ItemsEntity {
  text: string;
  checked: boolean;
}

export interface AttachmentsEntity {
  type: string;
  url: string;
}
