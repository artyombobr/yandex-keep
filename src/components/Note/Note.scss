@import "../../shared";

.reminder {
  display: flex;
  align-items: center;
  margin: 12px 26px;

  &__text {
    font-size: 18px;
    line-height: 24px;
    color: #fff;
    margin-left: 14px;
  }
}

.note {
  display: flex;
  flex-direction: column;
  border-radius: 16px;
  background-color: #fff;
  overflow: hidden;
  transition: box-shadow .4s;

  &__main {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding: 24px 0;
    flex-grow: 1;
    border-radius: 16px;
    width: 100%;
  }

  &__white {
    display: flex;
    flex-grow: 1;
    border-radius: 16px;
    background-color: #fff;
  }

  &__reminder {
    display: flex;
    flex-grow: 1;
    flex-direction: column;
    border-radius: 16px;
    background-color: #db5341;
  }

  &__attachment {
    display: flex;
    flex-grow: 1;
    flex-direction: column;
    justify-content: space-between;
    border-radius: 16px;
    background-color: #e7e7e7;
  }

  &__checklist {
    display: flex;
    flex-grow: 1;
    flex-direction: column;
    justify-content: space-between;
    border-radius: 16px;
    background-color: #fff;
  }

  &__image img {
    width: 100%;
    border-radius: 16px;
  }

  &__title {
    overflow: hidden;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    margin: 0 24px 12px;
    @include fontsize(22px, 24px, 24px, 28px, $small, $large);
    //-webkit-box-orient: vertical;
  }

  &__list {
    display: flex;
    width: fit-content;
    flex-direction: column;
    //margin: 0 24px 10px;
  }

  &__text {
    white-space: pre-line;
    margin: 0 24px 12px;
    @include fontsize(16px, 20px, 18px, 22px, $small, $large);
  }

  &__tags {
    display: flex;
    flex-wrap: wrap;
    overflow: hidden;
    text-overflow: ellipsis;
    margin: 0 24px;
  }

  &__tag {
    height: 24px;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    display: inline-block;
    line-height: 22px;
    font-size: 14px;
    margin-right: 12px;
    margin-bottom: 12px;
    padding: 0 12px;
    border-radius: 16px;
    background: rgba(51, 51, 51, .1);
    cursor: pointer;
    transition: background .4s;

    &:hover {
      background: rgba(51, 51, 51, .2);
    }
  }

  &__footer {
    flex-grow: 999;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
  }

  &__info {
    display: flex;
    padding: 0 24px 24px;
    justify-content: space-between;
  }

  &_active {
    position: fixed;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    z-index: 2;

    & .note__main {
      padding-top: 64px;
    }

    & .reminder {
      display: none;
    }
  }

  &:hover {
    box-shadow: 0 2px 6px rgba(197, 186, 186, .5);
  }
}

.site {
  display: flex;
  align-items: center;
  margin: 0 24px;
  padding: 12px 0;
  border-bottom: 1px solid white;

  &:last-child {
    border: none;
  }

  &__link {
    overflow: hidden;
    text-overflow: ellipsis;
    margin-left: 12px;
    font-size: 16px;
    line-height: 24px;
    text-decoration: none;
    font-weight: bold;
    transition: color .4s;

    &:hover {
      color: #db5341;
    }
  }
}

.image {
  width: 64px;
  height: 64px;
  border-radius: 4px;
  margin: 12px 12px;
}

.images {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 0 12px;

  &__icon {
    margin-left: 12px;
  }

  &__list {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-end;
  }
}

.icons {
  transition: opacity .4s;

  &__item {
    margin-right: 17px;
    cursor: pointer;
    opacity: .5;
    transition: opacity .4s;

    &:hover{
      opacity: 1;
    }

    & g{
      opacity: inherit;
    }
  }
}

.checklist {
  display: flex;
  flex-wrap: wrap;
  margin: 24px 24px 0;
}

.checkbox {
  position: relative;
  width: fit-content;
  margin-bottom: 12px;
  margin-right: 12px;

  & input {
    display: none;
  }

  &__span {
    line-height: 20px;
    font-size: 16px;
    cursor: pointer;
    @include fontsize(16px, 18px, 20px, 22px, $small, $large);

    &::before {
      content: "";
      display: inline-block;
      box-sizing: border-box;
      vertical-align: top;
      margin-right: 12px;
      border-radius: 3px;
      width: 18px;
      height: 18px;
      background-color: #D9D9D9;
    }

    &::after {
      content: "";
      display: none;
      position: absolute;
      top: 4px;
      left: 1px;
      width: 8px;
      height: 4px;
      border: solid 2px;
      border-right: none;
      border-top: none;
      transform: translate(3px, 0) rotate(-45deg);
    }
  }
}

input:checked ~ .checkbox__span {
  text-decoration: line-through;
}

input:checked ~ .checkbox__span::after {
  display: block;
}

input:checked ~ .checkbox__span::before {
  background-color: #FC0;
}

.l {
  grid-column-end: span 2;
  grid-row-end: span 2;
}

.m {
  grid-column-end: span 2;
}

.s {
  grid-column-end: span 2;
}

@media screen and (min-width: 768px){
  .icons{
    opacity: 0;
  }

  .note:hover .icons{
    opacity: 1;
  }

  .note__title {
    margin-bottom: 8px;
  }
}

@media screen and (min-width: 1024px) {
  .l {
    grid-column-end: span 4;
    grid-row-end: span 3; //заглушка
  }

  section.notes > div > div:nth-child(9) {
    grid-row-end: span 2; //заглушка
  }

  .m {
    grid-column-end: span 3;
  }

  .s {
    grid-column-end: span 2;
  }

  .note_active {
    height: auto;
    position: fixed;
    top: 25%;
    width: 50%;
    left: 25%;

    & .note__main {
      padding-top: 0;
    }

    & .reminder {
      display: block;
    }
  }
}
