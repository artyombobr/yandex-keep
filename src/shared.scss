$small: 375px;
$medium: 769px;
$large: 1366px;

@function strip-unit($number) {
  @return $number / ($number * 0 + 1);
}

@function calc-size($f-min, $f-max, $v-min, $v-max) {
  $f-min: strip-unit($f-min);
  $f-max: strip-unit($f-max);
  $v-min: strip-unit($v-min);
  $v-max: strip-unit($v-max);
  $k: ($f-max - $f-min)/($v-max - $v-min);
  $b: $f-min - $k * $v-min;
  $b: $b * 1px;
  @return calc( #{$k} * 100vw + #{$b} );
}

@mixin fontsize($f-min, $f-max, $l-min, $l-max, $w-min, $w-max) {
  font-size: $f-min;
  @if $l-min !=0 {
    line-height: $l-min;
  }
  @media (min-width: $w-min+1) {
    font-size: calc-size($f-min, $f-max, $w-min, $w-max);
    @if $l-min !=0 {
      line-height: calc-size($l-min, $l-max, $w-min, $w-max);
    }
  }
  @media (min-width: $w-max) {
    font-size: $f-max;
    @if $l-max !=0 {
      line-height: $l-max;
    }
  }
}
