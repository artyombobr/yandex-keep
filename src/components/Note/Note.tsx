import * as React from 'react';
import Reminder from '../Reminder/Reminder';
import Attachments from '../Attachments/Attachments';
import NoteText from '../NoteText/NoteText';
import NoteList from '../NoteList/NoteList';
import {DataEntity, ItemsEntity, NotesEntity} from '../../shared';
import './Note.scss';
import NoteImage from "../NoteImage/NoteImage";


function Note(props: { data: any }) {
  const {data } = props;
  const { type, title, text, attachments, color, items, tags, reminder, created, size } = props.data;

  // const NoteType: { [key: string]: any } = {
  //   ['text']: <NoteText data={data} note={note} />,
  //   ['list']: <NoteList data={data} note={note}/>,
  //   ['image']: <NoteImage data={data} note={note}/>
  // };

   return (
    <div className={'note ' + size}>
      <div className='note__reminder'>
        {reminder && <Reminder reminder={reminder} />}
        <div className='note__attachment'>
          {/*{NoteType[type]}*/}
          {attachments && <Attachments attachments={attachments} />}
        </div>
      </div>
    </div>
  );
}

export default Note;
