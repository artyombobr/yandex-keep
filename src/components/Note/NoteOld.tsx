import moment from 'moment';
import 'moment/locale/ru';
import * as React from 'react';
import { useState } from 'react';
import {DataEntity, ItemsEntity, NotesEntity} from '../../shared';
import './Note.scss';

import { ReactComponent as Check } from './svg/check.svg';
import { ReactComponent as Edit } from './svg/edit.svg';
import { ReactComponent as Picture } from './svg/picture.svg';
import { ReactComponent as Site } from './svg/site.svg';
import { ReactComponent as Warning } from './svg/warning.svg';

function NoteOld(props: { data: DataEntity; note: NotesEntity }) {
  const data = props.data;
  const { type, title, text, attachments, color, items, tags, reminder, created, size } = props.note;

  function SetBackground() {
    const opacity = 0.4;
    if (color !== undefined) {
      let hex: string = data.colors[color].color;
      hex = hex.replace('#', '');
      let r: number = parseInt(hex.substring(0, 2), 16);
      let g: number = parseInt(hex.substring(2, 4), 16);
      let b: number = parseInt(hex.substring(4, 6), 16);
      r = Math.round(r * opacity + 255 * (1 - opacity));
      g = Math.round(g * opacity + 255 * (1 - opacity));
      b = Math.round(b * opacity + 255 * (1 - opacity));
      return {backgroundColor: 'rgb(' + r + ',' + g + ',' + b + ')'};
    } else {
      return ({backgroundColor: '#fff'});
    }
  }

  function tagsList() {
    const list = [];
    if (tags !== undefined) {
      for (let i = 0; i < tags.length; i++) {
        list.push(<div className='note__tag' key={i}>{data.tags[tags[i]].tag}</div>);
      }
    }
    return list;
  }

  function linkCut(link: string) {
    const pos = link.indexOf('://');
    if (pos) {
      return link.substr(pos + 3);
    } else {
      return link;
    }
  }

  function attachmentsList() {
    const list = [];
    const listSites = [];
    const listImages = [];
    if (attachments !== undefined) {
      for (let i = 0; i < attachments.length; i++) {
        if (attachments[i].type === 'link') {
          listSites.push(
            <div className='site' key={i}>
              <Site />
              <a className='site__link' href={attachments[i].url}>{linkCut(attachments[i].url)}</a>
            </div>
          );
        } else {
          listImages.push(<img className='image' src={attachments[i].url} alt='Прикрепленное изображение' key={i} />);
        }
      }
    }
    if (listSites.length) { list.push(<div className='sites' key={0}>{listSites}</div>); }
    if (listImages.length) {
      list.push(
        <div className='images' key={1}>
          <Picture className='images__icon' />
          <div className='images__list'>{listImages}</div>
        </div>
      );
    }
    return list;
  }

  function reminderTime() {
    if (reminder !== undefined) {
      return (
        <div className='reminder'>
          <Warning />
          <div className='reminder__text'>{moment(reminder).fromNow()}</div>
        </div>
      );
    } else {
      return null;
    }
  }

  function renderText() {
    const [active, setActive] = useState(false);
    return (
      <div
        className={['note', size, active ? 'note_active' : ''].join(' ')}
        onClick={() => setActive(!active)}
      >
        <div className='note__reminder'>
          {reminderTime()}
          <div className='note__attachment'>
              <div className='note__main' style={SetBackground()}>
              <h1 className='note__title'>{title}</h1>
              <div className='note__text'>{text}</div>
              <div className='note__footer'>
                <div className='note__tags'>{tagsList()}</div>
                <div className='note__info'>
                  <div className='icons'>
                    <Check className='icons__item' />
                    <Edit className='icons__item' />
                  </div>
                  {moment(created).fromNow()}
                </div>
              </div>
            </div>
            {attachmentsList()}
          </div>
        </div>
      </div>
    );
  }

  function renderList() {
    let listChecked;
    let listNotChecked;
    if (items) {
      listChecked = items.map((item: ItemsEntity, index: number) => {
        if (item.checked) {
          return (
            <label className='checkbox' key={index}>
              <input type='checkbox' defaultChecked={true} />
              <span className='checkbox__span'>{item.text}</span>
            </label>
          );
        } else {
          return null;
        }
      });
      listNotChecked = items.map((item: ItemsEntity, index: number) => {
        if (!item.checked) {
          return (
            <label className='checkbox' key={index}>
              <input type='checkbox' />
              <span className='checkbox__span'>{item.text}</span>
            </label>
          );
        } else {
          return null;
        }
      });
    }
    const [active, setActive] = useState(false);
    return (
      <div
        className={['note', size, active ? 'note_active' : ''].join(' ')}
        onClick={() => setActive(!active)}
      >
        <div className='note__reminder'>
          {reminderTime()}
          <div className='note__attachment'>
            <div className='note__checklist'>
              <div className='note__main' style={SetBackground()}>
                <h1 className='note__title'>{title}</h1>
                <div className='note__list'>{listNotChecked}</div>
              </div>
              <div className='checklist'>{listChecked}</div>
              <div className='note__footer'>
                <div className='note__tags'>{tagsList()}</div>
                <div className='note__info'>
                  <div className='icons'>
                    <Check className='icons__item' />
                    <Edit className='icons__item' />
                  </div>
                  {moment(created).fromNow()}
                </div>
              </div>
            </div>
            {attachmentsList()}
          </div>
        </div>
      </div>
    );
  }

  function renderImage() {
    function image() {
      return (
        <div className='note__image'>
          <picture>
            <source media='(max-width: 320px)' srcSet='https://via.placeholder.com/320' />
            <source media='(max-width: 465px)' srcSet='https://via.placeholder.com/465' />
            <source media='(max-width: 650px)' srcSet='https://via.placeholder.com/650' />
            <img
              src='https://via.placeholder.com/1020'
              alt='Изображение'
            />
          </picture>
        </div>
      );
    }
    const [active, setActive] = useState(false);
    return (
      <div
        className={['note', size, active ? 'note_active' : ''].join(' ')}
        onClick={() => setActive(!active)}
      >
        <div className='note__reminder'>
          {reminderTime()}
          <div className='note__attachment'>
            <div className='note__main' style={SetBackground()}>
              {image()}
              <h1 className='note__title'>{title}</h1>
              <div className='note__text'>{text}</div>
              <div className='note__tags'>{tagsList()}</div>
              <div className='note__info'>
                <div className='icons'>
                  <Check className='icons__item' />
                  <Edit className='icons__item' />
                </div>
                {moment(created).fromNow()}
              </div>
            </div>
            {attachmentsList()}
          </div>
        </div>
      </div>
    );
  }

  let noteItem;

  switch (type) {
    case 'text':
      noteItem = renderText();
      break;
    case 'list':
      noteItem = renderList();
      break;
    case 'image':
      noteItem = renderImage();
      break;
    default:
      noteItem = renderText();
  }

  return (noteItem);
}

export default NoteOld;
