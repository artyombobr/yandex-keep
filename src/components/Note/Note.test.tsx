import React from 'react';
import { shallow } from 'enzyme';
import { Note } from './Note';
import { ColorsEntity } from '../../shared';

describe('Note', () => {
  const colors: ColorsEntity[] = [
    {
      id: 0,
      color: '#E84747',
    },
  ];
  const note = {
    type: 'text',
    title: 'Не забыть выгулять Сиба-Ину',
    color: 0,
    size: 's',
    created: 1520160803000,
  };
  it('renders correctly', () => {
    const component = shallow(<Note colors={colors} note={note} />);
    expect(component).toMatchSnapshot();
  });
});
