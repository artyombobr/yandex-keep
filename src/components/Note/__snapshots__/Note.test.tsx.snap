// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Note
    colors={
      Array [
        Object {
          "color": "#E84747",
          "id": 0,
        },
      ]
    }
    note={
      Object {
        "color": 0,
        "created": 1520160803000,
        "size": "s",
        "title": "Не забыть выгулять Сиба-Ину",
        "type": "text",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="note__reminder"
      >
        <div
          className="note__attachment"
        >
          <NoteText
            colors={
              Array [
                Object {
                  "color": "#E84747",
                  "id": 0,
                },
              ]
            }
            note={
              Object {
                "color": 0,
                "created": 1520160803000,
                "size": "s",
                "title": "Не забыть выгулять Сиба-Ину",
                "type": "text",
              }
            }
          />
        </div>
      </div>,
      "className": "note s",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <div
            className="note__attachment"
          >
            <NoteText
              colors={
                Array [
                  Object {
                    "color": "#E84747",
                    "id": 0,
                  },
                ]
              }
              note={
                Object {
                  "color": 0,
                  "created": 1520160803000,
                  "size": "s",
                  "title": "Не забыть выгулять Сиба-Ину",
                  "type": "text",
                }
              }
            />
          </div>,
        ],
        "className": "note__reminder",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <NoteText
                colors={
                  Array [
                    Object {
                      "color": "#E84747",
                      "id": 0,
                    },
                  ]
                }
                note={
                  Object {
                    "color": 0,
                    "created": 1520160803000,
                    "size": "s",
                    "title": "Не забыть выгулять Сиба-Ину",
                    "type": "text",
                  }
                }
              />,
              undefined,
            ],
            "className": "note__attachment",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "colors": Array [
                  Object {
                    "color": "#E84747",
                    "id": 0,
                  },
                ],
                "note": Object {
                  "color": 0,
                  "created": 1520160803000,
                  "size": "s",
                  "title": "Не забыть выгулять Сиба-Ину",
                  "type": "text",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            undefined,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="note__reminder"
        >
          <div
            className="note__attachment"
          >
            <NoteText
              colors={
                Array [
                  Object {
                    "color": "#E84747",
                    "id": 0,
                  },
                ]
              }
              note={
                Object {
                  "color": 0,
                  "created": 1520160803000,
                  "size": "s",
                  "title": "Не забыть выгулять Сиба-Ину",
                  "type": "text",
                }
              }
            />
          </div>
        </div>,
        "className": "note s",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <div
              className="note__attachment"
            >
              <NoteText
                colors={
                  Array [
                    Object {
                      "color": "#E84747",
                      "id": 0,
                    },
                  ]
                }
                note={
                  Object {
                    "color": 0,
                    "created": 1520160803000,
                    "size": "s",
                    "title": "Не забыть выгулять Сиба-Ину",
                    "type": "text",
                  }
                }
              />
            </div>,
          ],
          "className": "note__reminder",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <NoteText
                  colors={
                    Array [
                      Object {
                        "color": "#E84747",
                        "id": 0,
                      },
                    ]
                  }
                  note={
                    Object {
                      "color": 0,
                      "created": 1520160803000,
                      "size": "s",
                      "title": "Не забыть выгулять Сиба-Ину",
                      "type": "text",
                    }
                  }
                />,
                undefined,
              ],
              "className": "note__attachment",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "colors": Array [
                    Object {
                      "color": "#E84747",
                      "id": 0,
                    },
                  ],
                  "note": Object {
                    "color": 0,
                    "created": 1520160803000,
                    "size": "s",
                    "title": "Не забыть выгулять Сиба-Ину",
                    "type": "text",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              undefined,
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
